src/NuGet.Clients/NuGet.SolutionRestoreManager/VsSolutionRestoreService.cs

```diff
@@ -220,6 +223,29 @@ private static PackageSpec ToPackageSpec(ProjectNames projectNames, IVsProjectRe
             return packageSpec;
         }
 
+        private static NuGetVersion GetPackageVersion(IVsTargetFrameworks tfms)
+        {
+            // $(PackageVersion) property if set overrides the $(Version)
+            var versionPropertyValue = 
+                GetNonEvaluatedPropertyOrNull(tfms, PackageVersion) 
+                ?? GetNonEvaluatedPropertyOrNull(tfms, Version);
+
+            return versionPropertyValue != null
+                ? NuGetVersion.Parse(versionPropertyValue)
+                : PackageSpec.DefaultVersion;
+        }
+
+        // Trying to fetch a property value from tfm property bags.
+        // If defined the property should have identical values in all of the occurances.
+        private static string GetNonEvaluatedPropertyOrNull(IVsTargetFrameworks tfms, string propertyName)
+        {
+            return tfms
+                .Cast<IVsTargetFrameworkInfo>()
+                .Select(tfm => GetPropertyValueOrNull(tfm.Properties, propertyName))
+                .Distinct()
+                .SingleOrDefault();
```

Comment:
I would say this is the most correct, but will it be too difficult for the user to figure out if this silently returns null and 1.0.0 is used as the project version?